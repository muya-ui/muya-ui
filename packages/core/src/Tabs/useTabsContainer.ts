import { useEffect, useRef } from 'react';

import { useEventCallback } from '@muya-ui/utils';

import { ISwipeManager, useSwipe } from '../Swipe';
import { ITabsContainerProps } from './types';

export default function useTabsContainer(props: Pick<ITabsContainerProps, 'onChange'>) {
  const { onChange } = props;
  const { onPrev, onNext, onStepsChange, hasNext, hasPrev, stepIndex, manager } = useSwipe();
  const containerOffset = useRef(0);
  const updateContainerOffset = () => {
    if (manager) {
      containerOffset.current = manager.offset;
    }
  };
  const btnPrev = useEventCallback(() => {
    updateContainerOffset();
    onPrev();
  });
  const btnNext = useEventCallback(() => {
    updateContainerOffset();
    onNext();
  });

  const innerOnStepsChange = (manager: ISwipeManager) => {
    onStepsChange(manager);
    updateContainerOffset();
    onChange && onChange();
  };
  const onTransitionEnd = () => {
    updateContainerOffset();
    onChange && onChange();
  };
  useEffect(() => {
    if (manager) {
      const newOffset = manager.offset;
      if (newOffset !== containerOffset.current) {
        onChange && onChange(newOffset - containerOffset.current);
        containerOffset.current = newOffset;
      }
    }
  });
  return {
    onStepsChange: innerOnStepsChange,
    onPrev: btnPrev,
    onNext: btnNext,
    hasNext,
    hasPrev,
    stepIndex,
    onTransitionEnd,
  };
}
