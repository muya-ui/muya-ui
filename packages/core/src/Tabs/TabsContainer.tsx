import React from 'react';
import styled, { css } from 'styled-components';

import { LeftIcon, RightIcon } from '@muya-ui/theme-light';

import { IInlineButtonPureProps, InlineButton } from '../Button';
import Swipe from '../Swipe';
import { Omit } from '../types';
import addPx from '../utils/addPx';
import memoForwardRef from '../utils/memoForwardRef';
import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { ITabsContainerProps } from './types';
import useTabsContainer from './useTabsContainer';

const InnerButton = styled(InlineButton)``;

const StyledSwipe = styled(Swipe)`
  z-index: 1;
`;

type IButtonContainerProps = Omit<IInlineButtonPureProps, 'type'> &
  Pick<ITabsContainerProps, 'type'> &
  React.HTMLAttributes<HTMLDivElement>;

const BaseNode = styled.div``;

const ButtonContainerPure = memoForwardRef<HTMLDivElement, IButtonContainerProps>((props, ref) => {
  const {
    type,
    size,
    htmlType,
    disabled,
    busy,
    component: asProp,
    children,
    prefixNode,
    suffixNode,
    onClick,
    ...otherProps
  } = props;
  return (
    <BaseNode {...otherProps} ref={ref}>
      <InnerButton onClick={onClick} busy={busy}>
        {children}
      </InnerButton>
    </BaseNode>
  );
});

const ButtonContainer = styled(ButtonContainerPure)<IButtonContainerProps>`
  ${props => {
    const { theme, size = 'm', type } = props;
    const token = theme.components.Tabs;
    const height = type === 'line' ? token.height[size] : token.card.height[size];
    const padding = token.padding[size];
    let paddingCss;
    if (type === 'line' && padding) {
      paddingCss = css`
        padding: ${padding};
      `;
    }
    return css`
      & ${InnerButton} {
        width: ${height}px;
        min-height: ${addPx(height)};
        align-items: center;
        justify-content: center;
        ${paddingCss}
      }
    `;
  }}
`;

const TabsContainerPure = memoForwardRef<HTMLDivElement, ITabsContainerProps>((props, ref) => {
  const {
    children,
    onChange,
    height,
    size = 'm',
    type,
    swipe,
    // from swipe
    enableDiffChildren,
    equalNum,
    gutter,
    duration,
    ...otherProps
  } = props;
  const {
    onPrev,
    onNext,
    onStepsChange,
    hasNext,
    hasPrev,
    stepIndex,
    onTransitionEnd,
  } = useTabsContainer({
    onChange,
  });
  const theme = useTheme();
  const swipable = hasNext || hasPrev;
  const { Tabs: token } = theme.components;
  const tokenIcon = {
    prev: LeftIcon,
    next: RightIcon,
    ...token.icon,
  };
  const PrevIcon = tokenIcon.prev;
  const NextIcon = tokenIcon.next;
  const iconSize = token.iconSize[size];
  const iconStyle = {
    width: iconSize,
    height: iconSize,
  };
  const offSwipe = !swipe;

  const prevBtn =
    swipable && !offSwipe ? (
      <ButtonContainer size={size} theme={theme} type={type} onClick={onPrev} busy={!hasPrev}>
        <PrevIcon style={iconStyle} />
      </ButtonContainer>
    ) : null;
  const nextBtn =
    swipable && !offSwipe ? (
      <ButtonContainer size={size} theme={theme} type={type} onClick={onNext} busy={!hasNext}>
        <NextIcon style={iconStyle} />
      </ButtonContainer>
    ) : null;

  let innerChildren = children;
  if (!offSwipe) {
    const swipeProps = {
      enableDiffChildren,
      equalNum,
      gutter,
      duration,
    };
    innerChildren = (
      <StyledSwipe
        {...swipeProps}
        onTransitionEnd={onTransitionEnd}
        onStepsChange={onStepsChange}
        stepIndex={stepIndex}
      >
        {children}
      </StyledSwipe>
    );
  }

  return (
    <BaseNode {...otherProps} ref={ref}>
      {prevBtn}
      {innerChildren}
      {nextBtn}
    </BaseNode>
  );
});

const TabsContainer = styled(TabsContainerPure)`
  display: flex;
  flex-wrap: nowrap;
  white-space: nowrap;
  align-items: center;
  flex: 1;
  overflow-x: hidden;
  z-index: 1;
  ${props => {
    const { height } = props;
    if (height) {
      return css`
        height: ${height}px;
        align-items: center;
      `;
    }
  }}
`;

export default withThemeForStyled(TabsContainer);
