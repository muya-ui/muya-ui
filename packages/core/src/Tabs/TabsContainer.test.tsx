import React from 'react';
import renderer from 'react-test-renderer';
import mockHook from 'test/utils/mockHook';

import TabsContainer from './TabsContainer';
import * as useTabsContainer from './useTabsContainer';

test('测试 TabsContainer', () => {
  mockHook(useTabsContainer).returns({
    hasNext: true,
    hasPrev: true,
    stepIndex: 0,
  });
  const tree = renderer
    .create(
      <>
        <TabsContainer type="line"></TabsContainer>
        <TabsContainer type="card" size="l"></TabsContainer>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 TabsContainer swipe=true', () => {
  mockHook(useTabsContainer).returns({
    hasNext: true,
    hasPrev: true,
    stepIndex: 1,
  });
  const tree = renderer
    .create(
      <>
        <TabsContainer type="line" swipe></TabsContainer>
        <TabsContainer type="card" size="l" swipe></TabsContainer>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
