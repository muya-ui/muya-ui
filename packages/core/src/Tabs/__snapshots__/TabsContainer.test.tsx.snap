// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 TabsContainer 1`] = `
Array [
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  white-space: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-x: hidden;
  z-index: 1;
}

<div
    className="c0"
  />,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  white-space: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-x: hidden;
  z-index: 1;
}

<div
    className="c0"
  />,
]
`;

exports[`测试 TabsContainer swipe=true 1`] = `
Array [
  .c5 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c2 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c6 {
  position: relative;
  overflow-x: hidden;
}

.c8 {
  position: relative;
  -webkit-transition: -webkit-transform 450ms cubic-bezier(0.4,0,0.2,1);
  -webkit-transition: transform 450ms cubic-bezier(0.4,0,0.2,1);
  transition: transform 450ms cubic-bezier(0.4,0,0.2,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100000px;
  height: 100%;
}

.c4 {
  background: transparent;
  border: none;
  padding: 0;
  color: rgba(102,102,102,1);
  font-weight: 600;
  font-size: 14px;
  cursor: pointer;
}

.c4:hover {
  color: rgba(51,139,255,1);
}

.c4:active {
  color: rgba(12,108,235,1);
}

.c4 + .c3 {
  margin-left: 8px;
}

.c7 {
  z-index: 1;
}

.c1 .c3 {
  width: 32px;
  min-height: 32px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 0 8px 0;
}

.c9 .c3 {
  width: 38px;
  min-height: 38px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  white-space: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-x: hidden;
  z-index: 1;
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <button
        className="c2 c3 c4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="false"
          className="c5"
          focusable="false"
          style={
            Object {
              "height": 8,
              "width": 8,
            }
          }
          viewBox="0 0 1024 1024"
        >
          <path
            d="M746.667 191.733V834.54a21.333 21.333 0 01-33.259 17.685l-453.76-306.09a21.333 21.333 0 01-.779-34.817L712.64 174.613a21.333 21.333 0 0134.027 17.12z"
          />
        </svg>
      </button>
    </div>
    <div
      className="c6 c7"
    >
      <div
        className="c8 muya-swipe-panel"
        onTransitionEnd={[Function]}
        style={
          Object {
            "transform": "translate(0px, 0)",
          }
        }
      />
    </div>
    <div
      className="c1"
    >
      <button
        className="c2 c3 c4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="false"
          className="c5"
          focusable="false"
          style={
            Object {
              "height": 8,
              "width": 8,
            }
          }
          viewBox="0 0 1024 1024"
        >
          <path
            d="M298.667 191.733V834.54a21.333 21.333 0 0033.258 17.685l453.76-306.09a21.333 21.333 0 00.779-34.817l-453.77-336.704a21.333 21.333 0 00-34.027 17.12z"
          />
        </svg>
      </button>
    </div>
  </div>,
  .c5 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c2 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c6 {
  position: relative;
  overflow-x: hidden;
}

.c8 {
  position: relative;
  -webkit-transition: -webkit-transform 450ms cubic-bezier(0.4,0,0.2,1);
  -webkit-transition: transform 450ms cubic-bezier(0.4,0,0.2,1);
  transition: transform 450ms cubic-bezier(0.4,0,0.2,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100000px;
  height: 100%;
}

.c4 {
  background: transparent;
  border: none;
  padding: 0;
  color: rgba(102,102,102,1);
  font-weight: 600;
  font-size: 14px;
  cursor: pointer;
}

.c4:hover {
  color: rgba(51,139,255,1);
}

.c4:active {
  color: rgba(12,108,235,1);
}

.c4 + .c3 {
  margin-left: 8px;
}

.c7 {
  z-index: 1;
}

.c9 .c3 {
  width: 32px;
  min-height: 32px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 0 8px 0;
}

.c1 .c3 {
  width: 38px;
  min-height: 38px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  white-space: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-x: hidden;
  z-index: 1;
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <button
        className="c2 c3 c4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="false"
          className="c5"
          focusable="false"
          style={
            Object {
              "height": 12,
              "width": 12,
            }
          }
          viewBox="0 0 1024 1024"
        >
          <path
            d="M746.667 191.733V834.54a21.333 21.333 0 01-33.259 17.685l-453.76-306.09a21.333 21.333 0 01-.779-34.817L712.64 174.613a21.333 21.333 0 0134.027 17.12z"
          />
        </svg>
      </button>
    </div>
    <div
      className="c6 c7"
    >
      <div
        className="c8 muya-swipe-panel"
        onTransitionEnd={[Function]}
        style={
          Object {
            "transform": "translate(0px, 0)",
          }
        }
      />
    </div>
    <div
      className="c1"
    >
      <button
        className="c2 c3 c4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="false"
          className="c5"
          focusable="false"
          style={
            Object {
              "height": 12,
              "width": 12,
            }
          }
          viewBox="0 0 1024 1024"
        >
          <path
            d="M298.667 191.733V834.54a21.333 21.333 0 0033.258 17.685l453.76-306.09a21.333 21.333 0 00.779-34.817l-453.77-336.704a21.333 21.333 0 00-34.027 17.12z"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;
